app/src/main/java/org/thoughtcrime/securesms/util/ProfileUtil.java

```diff
@@ -0,0 +1,333 @@
+  }
+
+  public static Single<Pair<Recipient, ServiceResponse<ProfileAndCredential>>> retrieveProfile(@NonNull Context context,
+                                                                                               @NonNull Recipient recipient,
+                                                                                               @NonNull SignalServiceProfile.RequestType requestType,
+                                                                                               @NonNull ProfileService profileService)
+  {
+    return retrieveProfile(context, recipient, requestType, profileService, true);
+  }
+
+  private static Single<Pair<Recipient, ServiceResponse<ProfileAndCredential>>> retrieveProfile(@NonNull Context context,
+                                                                                                @NonNull Recipient recipient,
+                                                                                                @NonNull SignalServiceProfile.RequestType requestType,
+                                                                                                @NonNull ProfileService profileService,
+                                                                                                boolean allowUnidentifiedAccess)
+  {
+    Optional<UnidentifiedAccess> unidentifiedAccess = allowUnidentifiedAccess ? getUnidentifiedAccess(context, recipient) : Optional.absent();
+    Optional<ProfileKey>         profileKey         = ProfileKeyUtil.profileKeyOptional(recipient.getProfileKey());
+
+    return Single.fromCallable(() -> toSignalServiceAddress(context, recipient))
+                 .flatMap(address -> profileService.getProfile(address, profileKey, unidentifiedAccess, requestType, Locale.getDefault()).map(p -> new Pair<>(recipient, p)))
+                 .onErrorReturn(t -> new Pair<>(recipient, ServiceResponse.forUnknownError(t)));
+  }
+
+  public static @Nullable String decryptString(@NonNull ProfileKey profileKey, @Nullable byte[] encryptedString)
+      throws InvalidCiphertextException, IOException
+  {
+    if (encryptedString == null) {
+      return null;
+    }
+
+    ProfileCipher profileCipher = new ProfileCipher(profileKey);
+    return profileCipher.decryptString(encryptedString);
+  }
+
+  public static @Nullable String decryptString(@NonNull ProfileKey profileKey, @Nullable String encryptedStringBase64)
+      throws InvalidCiphertextException, IOException
+  {
+    if (encryptedStringBase64 == null) {
+      return null;
+    }
+
+    return decryptString(profileKey, Base64.decode(encryptedStringBase64));
+  }
+
+  @WorkerThread
+  public static @NonNull MobileCoinPublicAddress getAddressForRecipient(@NonNull Recipient recipient)
+      throws IOException, PaymentsAddressException
+  {
+    ProfileKey profileKey;
+    try {
+      profileKey = getProfileKey(recipient);
+    } catch (IOException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)
Similar issue at line numbers 162, 165.